src/base/bittorrent/tracker.h

```diff
@@ -49,21 +50,76 @@ namespace Http
 
 namespace BitTorrent
 {
-    struct Peer
+    class Peer
     {
-        QString ip;
-        QString peerId;
-        int port;
+    public:
+        Peer() = default;
+        Peer(QString&& ip, QString&& announce_ip, std::string&& peer_id, int port, QByteArray&& uid);
+
+        Peer(const Peer&) = default;
+        Peer(Peer&&) = default;
+        Peer& operator=(const Peer&) = default;
+        Peer& operator=(Peer&&) = default;
+
+        const QString& getIp() const;
 
         bool operator!=(const Peer &other) const;
         bool operator==(const Peer &other) const;
-        QString uid() const;
-        libtorrent::entry toEntry(bool noPeerId) const;
+        const QByteArray& uid() const;
+        // converts to appropriate entry taking into account whether 'clientIp' is a public ip address,  i.e. provides
+        // public ip if 'clientIP' belongs to a public network, otherwise gives connection ip
+        libtorrent::entry toEntryForClient(bool noPeerId, const QString& clientIp) const;
+
+    private:
+
+        // ip retrieved from connection
+        QString m_ip;
+        // ip obtained from announce query
+        QString m_announce_ip;
+        std::string m_peer_id;
+        int m_port;
+        QByteArray m_uid;
+
+        // if 'clientIP' belongs to a public network, returns public address found among connection and announce ip adrresses
+        const QString& getIpForClient(const QString& clientIP) const;
+        // percent-escapes given string
+        static std::string escape(const std::string& str);
+    };
+
+    class PeerBuilder
+    {
+    public:
+        Peer build();
+
+        void setIp(const QString& ip);
+        void setAnnounceIp(const QByteArray& ip);
+        void setPeerId(const QByteArray& id);
+        void setPort(int port);
+
+    private:
+        static const size_t PEER_ID_EXPECTED_LENGTH = 20;
+
+        // ip retrieved from connection
+        QString m_ip;
+        // ip obtained from announce query
+        QString m_announce_ip;
+        QByteArray m_peer_id;
+        int m_port;
+    };
+
+    struct IpUtil
```

Comment:
We don't allow pure static classes, so you must use namespaces in such cases.
But since here we have only module private functions, then just place it in an anonymous namespace in the .cpp file.